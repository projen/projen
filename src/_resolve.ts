import { types } from "util";
import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // Special resolution for few JavaScript built-in types
  // that by default would be stringified as empty objects ('{}')
  // as they are missing a `toJSON` implementation.
  switch (true) {
    case types.isRegExp(value):
      if (value.flags) {
        throw new Error(
          "RegExp with flags should be explicitly converted to a string"
        );
      }
      return value.source;

    case types.isSet(value):
      return resolve(Array.from(value), options);

    case types.isMap(value):
      return resolve(Object.fromEntries(value), options);

    case types.isBoxedPrimitive(value):
      return resolve(value.valueOf(), options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
