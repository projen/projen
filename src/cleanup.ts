import * as fs from "fs";
import * as path from "path";
import * as glob from "fast-glob";
import { PROJEN_DIR, PROJEN_MARKER } from "./common";
import * as logging from "./logging";

export const FILE_MANIFEST = `${PROJEN_DIR}/files.json`;

export function cleanup(dir: string, newFiles: string[], exclude: string[]) {
  try {
    const manifestFiles = getFilesFromManifest(dir);
    if (manifestFiles.length > 0) {
      // Use `FILE_MANIFEST` to remove files that are no longer managed by projen
      removeFiles(findOrphanedFiles(dir, manifestFiles, newFiles));
    } else {
      // Remove all files managed by projen with legacy logic
      removeFiles(findGeneratedFiles(dir, exclude));
    }
  } catch (e: any) {
    logging.warn(`warning: failed to clean up generated files: ${e.stack}`);
  }
}

function removeFiles(files: string[]) {
  for (const file of files) {
    fs.rmSync(file, { force: true, recursive: true });
  }
}

function findGeneratedFiles(dir: string, exclude: string[]) {
  const ignore = [
    ...readGitIgnore(dir),
    "node_modules/**",
    ...exclude,
    ".git/**",
  ];

  const files = glob.sync("**", {
    ignore,
    cwd: dir,
    dot: true,
    onlyFiles: true,
    followSymbolicLinks: false,
    absolute: true,
  });

  const generated = new Array<string>();

  for (const file of files) {
    const contents = fs.readFileSync(file, "utf-8");

    if (contents.includes(PROJEN_MARKER)) {
      generated.push(file);
    }
  }

  return generated;
}

function findOrphanedFiles(
  dir: string,
  oldFiles: string[],
  newFiles: string[]
) {
  return oldFiles
    .filter((old) => !newFiles.includes(old))
    .map((f: string) => path.resolve(dir, f));
}

function getFilesFromManifest(dir: string): string[] {
  try {
    const fileManifestPath = path.resolve(dir, FILE_MANIFEST);
    if (fs.existsSync(fileManifestPath)) {
      const fileManifest = JSON.parse(
        fs.readFileSync(fileManifestPath, "utf-8")
      );
      if (fileManifest.files) {
        return fileManifest.files;
      }
    }
  } catch (e: any) {
    logging.warn(
      `warning: unable to get files to clean from file manifest: ${e.stack}`
    );
  }

  return [];
}

function readGitIgnore(dir: string) {
  const filepath = path.join(dir, ".gitignore");
  if (!fs.existsSync(filepath)) {
    return [];
  }

  return fs
    .readFileSync(filepath, "utf-8")
    .split("\n")
    .filter((x) => x?.trim() !== "")
    .filter((x) => !x.startsWith("#") && !x.startsWith("!"))
    .map((x) => x.replace(/^\//, "")) // remove "/" prefix
    .map((x) => `${x}\n${x}/**`)
    .join("\n")
    .split("\n");
}
