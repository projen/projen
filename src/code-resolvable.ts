const CODE_RESOLVABLE_SYMBOL = Symbol.for('projen.CodeResolvable');

export interface IImportResolver {
  /**
   * Adds a named import from a module
   */  
  from(moduleName: string, importName: string, as?: string): ICodeResolvable;
}

/**
 * Interface for code that should be rendered as-is without quotes.
 */
export interface ICodeResolvable {
  /**
   * Renders the code as a string.
   * @returns The rendered code
   */
  render(): string;
  
  /**
   * Resolves imports needed by this code resolvable.
   * @param resolver - The IImportsResolver instance to resolve imports into
   */
  resolveImports?(resolver: IImportResolver): void;
}

/**
 * Abstract base class for code that should be rendered as-is without quotes.
 */
export abstract class CodeResolvable implements ICodeResolvable {
  /**
   * Checks if an object is a CodeResolvable instance.
   * @param obj - The object to check
   * @returns True if the object is CodeResolvable
   */
  public static isCodeResolvable(obj: any): obj is ICodeResolvable {
    return obj && typeof obj === 'object' && obj[CODE_RESOLVABLE_SYMBOL] === true;
  }

  /**
   * Creates a new CodeResolvable instance and marks it with the symbol.
   */
  public constructor() {
    // mark as CodeResolvable
    Object.defineProperty(this, CODE_RESOLVABLE_SYMBOL, { value: true });
  }

  /**
   * Renders the code as a string.
   * @returns The rendered code
   */
  public abstract render(): string;
}
