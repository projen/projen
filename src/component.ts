import { Construct, IConstruct } from "constructs";
import { Project } from "./project";
import {
  isComponent,
  findClosestProject,
  tagAsComponent,
} from "./util/constructs";

const autoIds = new WeakMap<IConstruct, number>();
const componentId = (scope: IConstruct) => {
  const nextId = (autoIds.get(scope) ?? 0) + 1;
  autoIds.set(scope, nextId);
  return `AutoId${nextId}`;
};

/**
 * Represents a project component.
 * @param project
 * @param id Unique id of the component. If not provided, an unstable AutoId is generated.
 */
export class Component extends Construct {
  /**
   * Test whether the given construct is a component.
   */
  public static isComponent(x: any): x is Component {
    return isComponent(x);
  }

  public readonly project: Project;

  constructor(scope: IConstruct, id?: string) {
    super(scope, id || `${new.target.name}#${componentId(scope)}`);
    tagAsComponent(this);
    this.node.addMetadata("type", "component");
    this.node.addMetadata("construct", new.target.name);

    this.project = findClosestProject(scope, new.target.name);
  }

  /**
   * Called before synthesis.
   */
  public preSynthesize() {}

  /**
   * Synthesizes files to the project output directory.
   */
  public synthesize() {}

  /**
   * Called after synthesis. Order is *not* guaranteed.
   */
  public postSynthesize() {}
}
