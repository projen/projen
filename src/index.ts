export * from "./ai-instructions";
export * from "./compare";
export * from "./component";
export * from "./dev-env";
export * from "./dependencies";
export * from "./docker-compose";
export * from "./dockerfile";
export * from "./file";
export * from "./gitattributes";
export * from "./gitpod";
export * from "./runner-options";
export * from "./ignore-file";
export * from "./ini";
export * from "./json";
export * from "./json-patch";
export * from "./logger";
export * from "./license";
export * from "./makefile";
export * from "./object-file";
export * from "./option-hints";
export * from "./project";
export * from "./project-build";
export * from "./project-tree";
export * from "./projects";
export * from "./projenrc";
export * from "./projenrc-json";
export * from "./readme";
export * from "./renovatebot";
export * from "./sample-file";
export * from "./semver";
export * from "./source-code";
export * from "./task";
export * from "./tasks";
export * from "./task-model";
export * from "./task-runtime";
export * from "./testing";
export * from "./textfile";
export * from "./toml";
export * from "./version";
export * from "./yaml";
export * from "./xmlfile";

// export submodules
export * as web from "./web";
export * as github from "./github";
export * as gitlab from "./gitlab";
export * as circleci from "./circleci";
export * as vscode from "./vscode";
export * as java from "./java";
export * as python from "./python";
export * as typescript from "./typescript";
export * as javascript from "./javascript";
export * as release from "./release";
export * as awscdk from "./awscdk";
export * as cdk8s from "./cdk8s";
export * as cdktf from "./cdktf";
export * as cdk from "./cdk";
export * as build from "./build";
