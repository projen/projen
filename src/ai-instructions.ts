import { Component } from "./component";
import { FileBase, IResolver } from "./file";
import { Project } from "./project";

/**
 * Supported AI coding assistants and their instruction file locations.
 */
export enum AiAgent {
  /**
   * GitHub Copilot - .github/copilot-instructions.md
   */
  GITHUB_COPILOT = "GitHub Copilot",

  /**
   * Cursor IDE - .cursor/rules/project.md
   */
  CURSOR = "Cursor",

  /**
   * Claude Code - CLAUDE.md
   */
  CLAUDE = "Claude",

  /**
   * Amazon Q - .amazonq/rules/project.md
   */
  AMAZON_Q = "Amazon Q",

  /**
   * Kiro - .kiro/steering/project.md
   */
  KIRO = "Kiro",
}

/**
 * Options for configuring AI tool instruction files.
 */
export interface AiInstructionsOptions {
  /**
   * Which AI agents to generate instruction files for.
   *
   * @default - All agents: [AiAgent.GITHUB_COPILOT, AiAgent.CURSOR, AiAgent.CLAUDE, AiAgent.AMAZON_Q]
   */
  readonly agents?: AiAgent[];

  /**
   * General instructions applicable to all agents.
   *
   * @default - no agent specific instructions
   */
  readonly instructions?: string[];

  /**
   * Per-agent custom instructions. Allows different instructions for different AI tools.
   *
   * @default - no agent specific instructions
   * @example
   * {
   *   [AiAgent.GITHUB_COPILOT]: {
   *     instructions: ["Use descriptive commit messages."]
   *   },
   *   [AiAgent.CURSOR]: {
   *     instructions: ["Prefer functional patterns.", "Always add tests."]
   *   }
   * }
   */
  readonly agentSpecificInstructions?: { [key: string]: string[] };

  /**
   * Include default instructions for projen and general best practices.
   *
   * Default instructions will only be included for agents provided in the `agents` option.
   * If `agents` is not provided, default instructions will be included for all agents.
   *
   * @default true
   */
  readonly includeDefaultInstructions?: boolean;
}

/**
 * Generates instruction files for AI coding assistants with projen-specific guidance.
 *
 * This component creates configuration files that help AI tools like GitHub Copilot,
 * Cursor IDE, Claude Code, and Amazon Q understand that the project is managed by projen
 * and should follow projen conventions.
 *
 * @example
 * const project = new TypeScriptProject({
 *   name: "my-project",
 *   defaultReleaseBranch: "main",
 * });
 *
 * // Basic usage - generates files for all supported AI agents
 * new AiInstructions(project);
 *
 * // Custom usage - specify which agents and add custom instructions
 * new AiInstructions(project, {
 *   agents: [AiAgent.GITHUB_COPILOT, AiAgent.CURSOR],
 *   agentSpecificInstructions: {
 *     [AiAgent.GITHUB_COPILOT]: {
 *       instructions: ["Always use descriptive commit messages."],
 *     },
 *   },
 * });
 *
 * // Add more instructions after instantiation
 * const ai = new AiInstructions(project);
 * ai.addInstructions("Use functional programming patterns.");
 * ai.addInstructions("Always write comprehensive tests.");
 */
export class AiInstructions extends Component {
  /**
   * Returns projen-specific instructions for AI agents.
   */
  public static projen(project: Project): string {
    return projenInstructions(project);
  }

  /**
   * Returns development best practices instructions for AI agents.
   */
  public static bestPractices(project: Project): string {
    return bestPracticesInstructions(project);
  }

  private readonly agents: AiAgent[];
  private readonly files: Map<AiAgent, AiInstructionsFile> = new Map();

  constructor(project: Project, options: AiInstructionsOptions = {}) {
    super(project);

    this.agents =
      options.agents ??
      (Object.values(AiAgent).filter(
        (v) => typeof v === "string"
      ) as AiAgent[]);

    // Assert files for declared agents
    for (const agent of this.agents) {
      this.assertInstructionsFile(agent);
    }

    if (options.includeDefaultInstructions ?? true) {
      this.addInstructions(
        AiInstructions.projen(project),
        AiInstructions.bestPractices(project)
      );
    }
  }

  /**
   * Create or return the instructions file.
   */
  private assertInstructionsFile(agent: AiAgent): AiInstructionsFile {
    if (this.files.has(agent)) {
      return this.files.get(agent)!;
    }

    const filePath = this.getAgentFilePath(agent);
    const file = new AiInstructionsFile(this.project, filePath, {
      committed: true,
      readonly: true,
    });
    this.files.set(agent, file);
    this.project.addPackageIgnore(file.path);
    return file;
  }

  /**
   * Adds instructions that will be included for all selected AI agents.
   *
   * @param instructions The instructions to add.
   * @example
   * aiInstructions.addInstructions("Always use TypeScript strict mode.");
   * aiInstructions.addInstructions("Prefer functional programming.", "Avoid mutations.");
   */
  public addInstructions(...instructions: string[]): void {
    for (const agent of this.files.keys()) {
      this.addAgentSpecificInstructions(agent, ...instructions);
    }
  }

  /**
   * Add instructions for a specific AI agent.
   *
   * This can also be used to add instructions for an AI agent that was previously not enabled.
   *
   * @param agent The AI agent to add instructions for
   * @param instructions The instruction(s) to add
   * @example
   * aiInstructions.addAgentSpecificInstructions(AiAgent.GITHUB_COPILOT, "Use descriptive commit messages.");
   */
  public addAgentSpecificInstructions(
    agent: AiAgent,
    ...instructions: string[]
  ): void {
    const file = this.assertInstructionsFile(agent);
    file.addInstructions(...instructions);
  }

  /**
   * Get the file path for a given AI agent.
   */
  private getAgentFilePath(agent: AiAgent): string {
    switch (agent) {
      case AiAgent.GITHUB_COPILOT:
        return ".github/copilot-instructions.md";
      case AiAgent.CURSOR:
        return ".cursor/rules/project.md";
      case AiAgent.CLAUDE:
        return "CLAUDE.md";
      case AiAgent.AMAZON_Q:
        return ".amazonq/rules/project.md";
      case AiAgent.KIRO:
        return ".kiro/steering/project.md";
      default:
        throw new Error(`Unknown AI agent: ${agent}`);
    }
  }
}

export class AiInstructionsFile extends FileBase {
  public readonly instructions: string[] = [];

  /**
   * Adds instructions to the instruction file.
   */
  public addInstructions(...instructions: string[]): void {
    this.instructions.push(...instructions);
  }

  protected synthesizeContent(resolver: IResolver): string | undefined {
    return resolver.resolve(this.instructions).join("\n\n") + "\n";
  }
}

function bestPracticesInstructions(project: Project): string {
  const projenCommand = project.projenCommand;
  return `# Development Best Practices

- **Always run build after changes**: After modifying any source or test file, run \`${projenCommand} build\` to ensure your changes compile and pass all tests.
- **Task completion criteria**: A task is not considered complete until:
  - All tests pass (\`${projenCommand} test\`)
  - There are no compilation errors (\`${projenCommand} compile\`)
  - There are no linting errors (usually part of the build, if not, run the linter defined in tasks.json)
  - The full build succeeds (\`${projenCommand} build\`)`;
}

function projenInstructions(project: Project): string {
  const projenCommand = project.projenCommand;
  return `# Projen-managed Project Instructions

This project is managed by [projen](https://github.com/projen/projen), a project configuration management tool.

## Important Guidelines

### Task Execution

- **Always use projen for task execution**: Run tasks using \`${projenCommand} <task-name>\` instead of directly using npm, yarn, or other package managers.
- **Check available tasks**: Look in \`.projen/tasks.json\` to see all available tasks, their descriptions, and steps.
- **Common tasks**:
  - \`${projenCommand}\` - Synthesize project configuration files
  - \`${projenCommand} build\` - Builds the project, including running test
  - \`${projenCommand} test\` - Runs tests only
  - \`${projenCommand} compile\` - Compiles the source code only

### File Modifications

- **DO NOT manually edit generated files**: Files marked with a comment like "~~ Generated by projen. To modify..." should never be edited directly.
- **Modify configuration in .projenrc**: To change project configuration, always edit the \`.projenrc.ts\`, \`.projenrc.py\` or \`.projenrc.json\` etc. file and then run \`${projenCommand}\` to regenerate the project files.
- **Check .projenrc first**: Before suggesting changes to package.json, tsconfig.json, or other configuration files, always check if these are managed by projen and suggest changes to .projenrc instead.

### Dependencies

- **Add dependencies through projen**: Use the projen configuration to add dependencies instead of manually editing package.json or using npm/yarn install directly.
- **Example**: In .projenrc, use methods like \`addDeps()\`, \`addDevDeps()\`, or \`addPeerDeps()\` to add dependencies.

### Workflow

1. Make changes to .projenrc configuration file
2. Run \`${projenCommand}\` to synthesize and update generated files
3. Review the changes
4. Commit both .projenrc and the generated files

## Projen Configuration

This project's configuration is defined in the .projenrc file at the root of the repository. All project metadata, dependencies, scripts, and tooling configuration should be managed through this file.

## Additional Resources

- [Projen Documentation](https://projen.io)
- [Projen GitHub Repository](https://github.com/projen/projen)`;
}
