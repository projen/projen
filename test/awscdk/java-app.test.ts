import * as xml from "xmlbuilder2";
import { awscdk } from "../../src";
import { Testing } from "../../src/testing";

test("happy flow", () => {
  const p = new awscdk.AwsCdkJavaApp({
    artifactId: "my-app",
    groupId: "org.acme",
    cdkVersion: "1.130.0",
    mainClass: "org.bacme.BoomApp",
    name: "my-app",
    version: "0.1.0",
  });

  const snapshot = Testing.synth(p);
  expect(Object.keys(snapshot).sort()).toStrictEqual([
    ".gitattributes",
    ".github/workflows/pull-request-lint.yml",
    ".gitignore",
    ".projen/deps.json",
    ".projen/files.json",
    ".projen/tasks.json",
    "README.md",
    "cdk.json",
    "pom.xml",
    "src/main/java/org/bacme/BoomApp.java",
    "src/test/java/org/bacme/MyTest.java",
  ]);
  expect(snapshot["cdk.json"]).toMatchSnapshot();
  expect(snapshot["pom.xml"]).toMatchSnapshot();
  expect(snapshot["src/test/java/org/bacme/MyTest.java"]).toMatchSnapshot();
  expect(snapshot["cdk.json"].app).toStrictEqual(
    "mvn exec:java --quiet -Dexec.mainClass=org.bacme.BoomApp"
  );
});

test("mainClass", () => {
  const p = new awscdk.AwsCdkJavaApp({
    artifactId: "my-app",
    groupId: "org.acme",
    cdkVersion: "1.130.0",
    mainClass: "org.acme.jojo.MyApp",
    name: "my-app",
    version: "0.1.0",
  });

  const snapshot = Testing.synth(p);
  expect(snapshot["cdk.json"].app).toStrictEqual(
    "mvn exec:java --quiet -Dexec.mainClass=org.acme.jojo.MyApp"
  );
});

test("deps", () => {
  const p = new awscdk.AwsCdkJavaApp({
    artifactId: "my-app",
    groupId: "org.acme",
    cdkVersion: "1.120.0",
    mainClass: "org.acme.jojo.MyApp",
    name: "my-app",
    version: "0.1.0",
    cdkDependencies: [
      "software.amazon.awscdk/aws-lambda",
      "software.amazon.awscdk/aws-sns",
    ],
  });

  p.addCdkDependency("software.amazon.awscdk/aws-sqs");

  const snapshot = Testing.synth(p);
  const pom = JSON.parse(
    xml.convert(snapshot["pom.xml"], {
      format: "json",
    })
  );

  expect(
    pom.project.dependencies.dependency.filter(
      (d: any) => d.groupId === "software.amazon.awscdk"
    )
  ).toStrictEqual([
    {
      artifactId: "aws-lambda",
      groupId: "software.amazon.awscdk",
      version: "[1.120.0,2.0.0)",
    },
    {
      artifactId: "aws-sns",
      groupId: "software.amazon.awscdk",
      version: "[1.120.0,2.0.0)",
    },
    {
      artifactId: "aws-sqs",
      groupId: "software.amazon.awscdk",
      version: "[1.120.0,2.0.0)",
    },
    {
      artifactId: "core",
      groupId: "software.amazon.awscdk",
      version: "[1.120.0,2.0.0)",
    },
    {
      artifactId: "assertions",
      groupId: "software.amazon.awscdk",
      version: "[1.120.0,2.0.0)",
      scope: "test",
    },
  ]);

  expect(
    pom.project.dependencies.dependency.filter(
      (d: any) => d.groupId === "software.constructs"
    )
  ).toStrictEqual([
    {
      artifactId: "constructs",
      groupId: "software.constructs",
      version: "[3.2.27,4.0.0)",
    },
  ]);
});

test("deps cdkv2", () => {
  const p = new awscdk.AwsCdkJavaApp({
    artifactId: "my-app",
    groupId: "org.acme",
    cdkVersion: "2.3.0",
    mainClass: "org.acme.jojo.MyApp",
    name: "my-app",
    version: "0.1.0",
  });

  const snapshot = Testing.synth(p);
  const pom = JSON.parse(
    xml.convert(snapshot["pom.xml"], {
      format: "json",
    })
  );

  expect(
    pom.project.dependencies.dependency.filter(
      (d: any) => d.groupId === "software.amazon.awscdk"
    )
  ).toStrictEqual([
    {
      artifactId: "aws-cdk-lib",
      groupId: "software.amazon.awscdk",
      version: "[2.3.0,3.0.0)",
    },
  ]);

  expect(
    pom.project.dependencies.dependency.filter(
      (d: any) => d.groupId === "software.constructs"
    )
  ).toStrictEqual([
    {
      artifactId: "constructs",
      groupId: "software.constructs",
      version: "[10.0.5,11.0.0)",
    },
  ]);
});

test("can disable sample code", () => {
  const p = new awscdk.AwsCdkJavaApp({
    artifactId: "my-app",
    groupId: "org.acme",
    cdkVersion: "1.120.0",
    mainClass: "org.acme.jojo.MyApp",
    sampleJavaPackage: "org.acme",
    sample: false,
    name: "my-app",
    version: "0.1.0",
  });

  const snapshot = Testing.synth(p);

  expect(Object.keys(snapshot)).not.toContain(
    "src/main/java/org/acme/jojo/MyApp.java"
  );
});
