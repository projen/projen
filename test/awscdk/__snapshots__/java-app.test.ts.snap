// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`happy flow 1`] = `
{
  "//": "~~ Generated by projen. To modify, edit src/test/java/projenrc.java and run "npx projen".",
  "app": "mvn exec:java --quiet -Dexec.mainClass=org.bacme.BoomApp",
  "output": "cdk.out",
}
`;

exports[`happy flow 2`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <modelVersion>4.0.0</modelVersion>
    <groupId>org.acme</groupId>
    <artifactId>my-app</artifactId>
    <version>0.1.0</version>
    <packaging>jar</packaging>
    <name>my-app</name>
    <properties>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    </properties>
    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.8.1</version>
                <configuration>
                    <source>1.8</source>
                    <target>1.8</target>
                </configuration>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-enforcer-plugin</artifactId>
                <version>3.0.0-M3</version>
                <configuration>
                    <rules>
                        <requireMavenVersion>
                            <version>3.6</version>
                        </requireMavenVersion>
                    </rules>
                </configuration>
                <executions>
                    <execution>
                        <id>enforce-maven</id>
                        <goals>
                            <goal>enforce</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-jar-plugin</artifactId>
                <version>3.2.0</version>
                <configuration>
                    <archive>
                        <index>true</index>
                        <manifest>
                            <addDefaultImplementationEntries>true</addDefaultImplementationEntries>
                            <addDefaultSpecificationEntries>true</addDefaultSpecificationEntries>
                        </manifest>
                    </archive>
                </configuration>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-javadoc-plugin</artifactId>
                <version>3.2.0</version>
                <configuration>
                    <failOnError>false</failOnError>
                    <show>protected</show>
                    <detectJavaApiLink>false</detectJavaApiLink>
                    <additionalJOptions>
                        <additionalJOption>-J-XX:+TieredCompilation</additionalJOption>
                        <additionalJOption>-J-XX:TieredStopAtLevel=1</additionalJOption>
                    </additionalJOptions>
                </configuration>
                <executions>
                    <execution>
                        <id>attach-javadocs</id>
                        <goals>
                            <goal>jar</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-source-plugin</artifactId>
                <version>3.2.1</version>
                <executions>
                    <execution>
                        <id>attach-sources</id>
                        <goals>
                            <goal>jar</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>org.codehaus.mojo</groupId>
                <artifactId>exec-maven-plugin</artifactId>
                <version>3.0.0</version>
            </plugin>
        </plugins>
    </build>
    <dependencies>
        <dependency>
            <groupId>software.amazon.awscdk</groupId>
            <artifactId>core</artifactId>
            <version>[1.130.0,2.0.0)</version>
        </dependency>
        <dependency>
            <groupId>software.constructs</groupId>
            <artifactId>constructs</artifactId>
            <version>[3.2.27,4.0.0)</version>
        </dependency>
        <dependency>
            <groupId>io.github.cdklabs</groupId>
            <artifactId>projen</artifactId>
            <version>99.99.99</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-api</artifactId>
            <version>5.7.0</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-engine</artifactId>
            <version>5.7.0</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>software.amazon.awscdk</groupId>
            <artifactId>assertions</artifactId>
            <version>[1.130.0,2.0.0)</version>
            <scope>test</scope>
        </dependency>
    </dependencies>
</project>
<!-- ~~ Generated by projen. To modify, edit src/test/java/projenrc.java and run "npx projen". -->"
`;

exports[`happy flow 3`] = `
"package org.bacme;

import org.junit.jupiter.api.Test;

public class MyTest {
  @Test
  public void testHello() {
    System.out.println("Hello, world!");
  }
}"
`;
